lib/Mojo/WebSocketProxy/CallingEngine.pm

```diff
@@ -0,0 +1,253 @@
+package Mojo::WebSocketProxy::CallingEngine;
+
+use strict;
+use warnings;
+
+use MojoX::JSON::RPC::Client;
+use Guard;
+use JSON;
+use Data::UUID;
+
+sub make_call_params {
+    my ($c, $req_storage) = @_;
+
+    my $args           = $req_storage->{args};
+    my $stash_params   = $req_storage->{stash_params};
+    my $require_auth   = $req_storage->{require_auth};
+    my $call_params_cb = delete $req_storage->{make_call_params};
+
+    my $call_params = $req_storage->{call_params};
+    $call_params->{args} = $args;
+
+    if (defined $stash_params) {
+        $call_params->{$_} = $c->stash($_) for @$stash_params;
+    }
+    if (defined $call_params_cb) {
+        my $cb_params = $call_params_cb->($c, $args);
+        $call_params->{$_} = $cb_params->{$_} for keys %$cb_params;
+    }
+
+    return $call_params;
+}
+
+sub get_rpc_response_cb {
+    my ($c, $req_storage) = @_;
+
+    my $success_handler = delete $req_storage->{success};
+    my $error_handler   = delete $req_storage->{error};
+    my $msg_type        = $req_storage->{msg_type};
+
+    if (my $rpc_response_cb = delete $req_storage->{rpc_response_cb}) {
+        return sub {
+            my $rpc_response = shift;
+            return $rpc_response_cb->($c, $req_storage->{args}, $rpc_response);
+        };
+    } else {
+        return sub {
+            my $rpc_response = shift;
+            if (ref($rpc_response) eq 'HASH' and exists $rpc_response->{error}) {
+                $error_handler->($c, $rpc_response, $req_storage) if defined $error_handler;
+                return error_api_response($c, $rpc_response, $req_storage);
+            } else {
+                $success_handler->($c, $rpc_response, $req_storage) if defined $success_handler;
+                store_response($c, $rpc_response);
+                return success_api_response($c, $rpc_response, $req_storage);
```

Comment:
Your suggestion will lead to duplicate code if we want to do something additional action when success is returned and nothing change response. E.g. we have to write code to make standard response.